src/libmeasurement_kit/ooni/whatsapp.cpp

```diff
@@ -0,0 +1,575 @@
+{ "2a03:2880:f211:c5:face:b00c::167/128" },
+{ "2a03:2880:f212:c5:face:b00c::167/128" },
+{ "2a03:2880:f213:c5:face:b00c::167/128" },
+{ "2a03:2880:f213:80c5:face:b00c::167/128" },
+{ "2a03:2880:f214:c5:face:b00c::167/128" },
+{ "2a03:2880:f215:c5:face:b00c::167/128" },
+{ "2a03:2880:f216:c5:face:b00c::167/128" },
+{ "2a03:2880:f217:c5:face:b00c::167/128" },
+{ "2a03:2880:f218:c3:face:b00c::167/128" },
+{ "2a03:2880:f219:c5:face:b00c::167/128" },
+{ "2a03:2880:f21a:c5:face:b00c::167/128" },
+{ "2a03:2880:f21b:c5:face:b00c::167/128" },
+{ "2a03:2880:f21c:c5:face:b00c::167/128" },
+{ "2a03:2880:f21c:80c5:face:b00c::167/128" },
+{ "2a03:2880:f21f:c5:face:b00c::167/128" },
+{ "2a03:2880:f221:c5:face:b00c::167/128" },
+{ "2a03:2880:f222:c5:face:b00c::167/128" },
+{ "2a03:2880:f223:c5:face:b00c::167/128" },
+{ "2a03:2880:f225:c4:face:b00c::167/128" },
+{ "2a03:2880:f226:c6:face:b00c::167/128" },
+{ "2a03:2880:f227:c5:face:b00c::167/128" } };
+
+
+std::vector<uint8_t> ip_to_bytes(std::string ip) {
+    sockaddr_storage ip_storage;
+    socklen_t ip_sz;
+    memset(&ip_storage, 0, sizeof ip_storage);
+
+    Error err = mk::net::make_sockaddr(ip, 0 /*port*/, &ip_storage, &ip_sz);
+    if (!!err) {
+        return {};
+    }
+
+    std::vector<uint8_t> bs;
+    uint32_t bsv4;
+    uint8_t *bsv6;
+    if (ip_sz == 16) { /* IPv4 */
+        bsv4 = ((struct sockaddr_in*)&ip_storage)->sin_addr.s_addr;
+        for (int i=0; i<4; i++) {
+            bs.insert(bs.end(), (bsv4 & 0xff));
+            bsv4 >>= 8;
+        }
+    } else if (ip_sz == 28) { /* IPv6 */
+        bsv6 = ((struct sockaddr_in6*)&ip_storage)->sin6_addr.s6_addr;
+        for (int i=0; i<16; i++) {
+            bs.insert(bs.end(), bsv6[i]);
+        }
+    }
+
+    return bs;
+}
+
+ErrorOr<bool> same_pre(std::vector<uint8_t> ip1, std::vector<uint8_t> ip2, int pre_bits) {
+    if (ip1.size() != ip2.size()) {
+        return GenericError();
+    }
+    // TODO: check pre_bits is sane
+    // full prefix bytes
+    int i = 0;
+    for (; i<(pre_bits/8); i++) {
+        if (ip1[i] != ip2[i]) {
+            return false;
+        }
+    }
+
+    int rem_bits = pre_bits % 8;
+    if (    (ip1[i] >> (8-rem_bits)) != \
+            (ip2[i] >> (8-rem_bits))) {
+        return false;
+    }
+    return true;
+}
+
+ErrorOr<bool> ip_in_net(std::string ip1, std::string ip_w_mask) {
+    auto ip1bs = ip_to_bytes(ip1);
+    for (auto b : ip1bs) {
+    }
+    auto ip2 = mk::split<std::vector<std::string>>(ip_w_mask, "/");
+    auto ip2bs = ip_to_bytes(ip2[0]);
+    for (auto b : ip2bs) {
+    }
+    auto pre_bits = std::stoi(ip2[1]);
+    return same_pre(ip1bs, ip2bs, pre_bits);
+}
+
+bool ip_in_nets(std::string ip, std::vector<std::string> nets) {
+    for (auto net : nets) {
+        auto res = ip_in_net(ip, net);
+        if (!!res && (res.as_value() == true)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+// XXX not even beginning to fill in entry
```

Comment:
@joelanders is this comment still actual; if so, can you fill the entry?

Reply:
acked